diam/sm/smparser/app.go

```diff
@@ -19,9 +19,34 @@ type Application struct {
 	id                          []uint32 // List of supported application IDs.
 }
 
+// SupportedApp holds properties of each locally supported App
+type SupportedApp struct {
+	ID      uint32
+	AppType string
+	Vendor  uint32
+}
+
+// PrepareSupportedApps prepares a list of locally supported apps
+func PrepareSupportedApps() []*SupportedApp {
+	locallySupportedApps := []*SupportedApp{}
+	for _, app := range dict.Default.Apps() {
+		if app.ID == 0 {
+			continue
+		}
+		addApp := new(SupportedApp)
+		addApp.ID = app.ID
+		addApp.AppType = app.Type
+		for _, vendor := range app.Vendor {
+			addApp.Vendor = vendor.ID
+		}
+		locallySupportedApps = append(locallySupportedApps, addApp)
+	}
+	return locallySupportedApps
+}
+
 // Parse ensures all acct or auth applications in the CE
 // exist in this server's dictionary.
-func (app *Application) Parse(d *dict.Parser) (failedAVP *diam.AVP, err error) {
+func (app *Application) Parse(d *dict.Parser, isServer uint8) (failedAVP *diam.AVP, err error) {
```

Comment:
i don't like this uint8 param here, because all calls to `Parse` look confusing now... instead, we should have a custom type and enums (iota) defining the supported types, client or server

```
type Role uint8

const (
  Server Role = iota + 1
  Client
)
```

Something like that..